### ISSUE TITLE
Step becomes a "go" in .NET 9

### ISSUE THREAD

**Issue by gregg-miskelly** [dotnet/runtime#109785]
### Description

.NET 9 contains a regression where ICorDebugStepper's are not always firing. The issue is non-deterministic, and causes intermittent failures in our debugging tests, and we have had a few customers reports of the problem.

### Reproduction Steps

A Microsoft employee can reproduce this problem by:

1. Enlisting in the 'Concord' repo
2. Following instructs in '<RepoRoot>/glass.md' for setting up a local test environment
3. Open Test Explorer
3. Enter “CSAsyncStepOverinTryBlock-NetCore-Latest-x64” in the filter box
4. Navigate to the test
5. Right click and invoke “Run Until Failure”

### Expected behavior

Stepper should complete

### Actual behavior

Stepper sometimes fails to detect when it is complete and this the debugger behaves as if there was no stepper

### Regression?

This regressed between .NET 8 and .NET 9

### Known Workarounds

None

### Configuration

So far all of the reports are from Windows, but I don't know if this is actually a causal connection.
We have at least mostly seen this on x64, but again, I am not sure if that is actually a causal connection or just because that is the most frequent configuration

**Comment by f2bo** [comment]
I'm seeing this all the time. I hadn't seen this issue before. I reported the problem here a few days ago: https://developercommunity.visualstudio.com/t/Single-stepping-is-broken-when-targettin/10802855.

I can reproduce it with the following simple steps:

- Create a new Console App project that targets .NET 8.
- Replace the contents of Program.cs with the following code:
  ``` C#
  foreach (Range range in new Range[5])
  {
     Console.WriteLine($"{range.Start}, {range.End}");
  }
  ```
- Execute in the debugger by using `Step Over` (**F10**) or `Step Into` (**F11**) to verify that the debugger is able to single-step through each line in the program correctly.

- Change the project file to target .NET 9.

- Execute again with `Step Over` (**F10**) or `Step Into` (**F11**) . Notice that single-stepping no longer advances to the next line as expected, but that execution suddenly continues without breaking, as if **F5** had been pressed.

NOTE: This behavior is erratic, so you may need to repeat a few times if it seems to work correctly.

**Comment by geoffrey-gaspari** [comment]
Same problem here.

**Comment by thaystg** [comment]
Hi @f2bo!
Every time that you run this app this "resume" happens?
It happens any time in the foreach loop or it's always in the last loop for example?
Can you share a video with the bug happening?

**Comment by f2bo** [comment]
Hi @thaystg,
> Every time that you run this app this "resume" happens?

Almost every time (see video below).

> It happens any time in the foreach loop or it's always in the last loop for example?

In this example, it usually happens on the first iteration of the foreach loop. I should mention that I've had this problem in other code and it's not always associated with a loop. In fact, in one of my tests, I saw it happen on a `Console.WriteLine` that I temporarily added before the loop using the code above.

> Can you share a video with the bug happening?

Sure. [Here](https://github.com/user-attachments/assets/5ee1a610-d310-484d-aa0e-8b0cefb3ecef) it is. 

The video shows the program being executed with **F10**, 3 times in succession. The first time, it run through the 5 iterations without failing, though I have to say that this is much less frequent. The remaining 2 times, instead of single-stepping, it resumes execution on the first iteration.

**Comment by f2bo** [comment]
Interesting fact. After disabling CET in the project file, I have been unable to reproduce the problem so far, at least in this simple example.

``` XML
<CETCompat>false</CETCompat>
```

**Comment by tommcdon** [comment]
Reactivated for backport to 9.0

**Comment by IamIC** [comment]
Same problem here.

**Comment by Ollhax** [comment]
I'm getting this all the time, both on my main machine and laptop.

**Comment by tommcdon** [comment]
Fixed via #110533

**Comment by tommcdon** [comment]
The fix for this issue has shipped in 9.0.2 - available from https://dotnet.microsoft.com/en-us/download/dotnet/9.0.
